core/filterconstraint.cpp

```diff
@@ -0,0 +1,939 @@
+{
+	const type_description *desc = get_type_description(type);
+	return desc ? desc - type_descriptions : -1;
+}
+
+extern "C"
+int filter_constraint_string_mode_to_index(enum filter_constraint_string_mode mode)
+{
+	const string_mode_description *desc = get_string_mode_description(mode);
+	return desc ? desc - string_mode_descriptions : -1;
+}
+
+extern "C"
+int filter_constraint_range_mode_to_index(enum filter_constraint_range_mode mode)
+{
+	const range_mode_description *desc = get_range_mode_description(mode);
+	return desc ? desc - range_mode_descriptions : -1;
+}
+
+extern "C"
+enum filter_constraint_type filter_constraint_type_from_index(int index)
+{
+	if (index >= 0 && index < (int)type_descriptions_count())
+		return type_descriptions[index].type;
+	return (enum filter_constraint_type)-1;
+}
+
+extern "C"
+enum filter_constraint_string_mode filter_constraint_string_mode_from_index(int index)
+{
+	if (index >= 0 && index < (int)string_mode_descriptions_count())
+		return string_mode_descriptions[index].mode;
+	return (enum filter_constraint_string_mode)-1;
+}
+
+extern "C"
+enum filter_constraint_range_mode filter_constraint_range_mode_from_index(int index)
+{
+	if (index >= 0 && index < (int)range_mode_descriptions_count())
+		return range_mode_descriptions[index].mode;
+	return (enum filter_constraint_range_mode)-1;
+}
+
+QString filter_constraint_type_to_string_translated(enum filter_constraint_type type)
+{
+	const type_description *desc = get_type_description(type);
+	return desc ? gettextFromC::tr(desc->translated) : "unknown";
+}
+
+QString filter_constraint_string_mode_to_string_translated(enum filter_constraint_string_mode type)
+{
+	const string_mode_description *desc = get_string_mode_description(type);
+	return desc ? gettextFromC::tr(desc->translated) : "unknown";
+}
+
+QString filter_constraint_range_mode_to_string_translated(enum filter_constraint_range_mode type)
+{
+	const range_mode_description *desc = get_range_mode_description(type);
+	return desc ? gettextFromC::tr(desc->translated) : "unknown";
+}
+
+QString filter_constraint_negate_to_string_translated(bool negate)
+{
+	return gettextFromC::tr(negate_description[negate]);
+}
+
+static enum filter_constraint_units type_to_unit(enum filter_constraint_type type)
+{
+	const type_description *desc = get_type_description(type);
+	return desc ? desc->units : FILTER_CONSTRAINT_NO_UNIT;
+}
+
+QString filter_constraint_get_unit(enum filter_constraint_type type)
+{
+	switch (type_to_unit(type)) {
+	case FILTER_CONSTRAINT_NO_UNIT:
+	default:
+		return QString();
+	case FILTER_CONSTRAINT_LENGTH_UNIT:
+		return get_depth_unit();
+	case FILTER_CONSTRAINT_TEMPERATURE_UNIT:
+		return get_temp_unit();
+	case FILTER_CONSTRAINT_WEIGHT_UNIT:
+		return get_weight_unit();
+	case FILTER_CONSTRAINT_VOLUMETRIC_FLOW_UNIT:
+		return get_volume_unit();
+	}
+}
+
+static int float_to_int(double f, enum filter_constraint_type type)
+{
```

Comment:
this is a fascinating function. I'd claim that the name is a bit misleading. `display_to_base_unit()` would do a better job conveying what it does, right? Yes, I get the unit-less case, but still...

Reply:
done